ci_tutorial/products_pipeline.rst

```diff
@@ -0,0 +1,103 @@
+value for users of those artifacts. In this example we will consider ``game/1.0`` and ``mapviewer/1.0`` the "products". Note that it is 
+possible to define different versions of the same package as products, for example, if we had to maintain different versions of the ``game`` for
+different customers, we could have ``game/1.0`` and ``game/2.3`` as well as different versions of ``mapviewer`` as products.
+
+The "products" approach, besides the advantage of focusing on the business value, has another very important advantage: it avoids having to model
+the dependency graph at the CI layer. It is a frequent attempt trying to model the inverse dependency model, that is, representing at the CI level
+the dependants or consumers of a given package. In our example, if we had configured a job for building the ``ai`` package, we could have another
+job for the ``engine`` package, that is triggered after the ``ai`` one, configuring such topology somehow in the CI system.
+
+But this approach does not scale at all and have very important limitations:
+  
+- The example above is relatively simple, but in practice dependency graphs can have many more packages, even several hundrends, making it very tedious and error prone to define all dependencies among packages in the CI
+- Dependencies evolve over time, and new versions are used, some dependencies are removed and newer dependencies are added. The simple relationship between repositories modeled at the CI level can result in a very inefficient, slow and time consuming CI, if not a fragile one that continuously breaks because some dependencies change.
+- The combinatorial nature that happens downstream a dependency graph, where a relatively stable top dependency, lets say ``mathlib/1.0`` might be used by multiple consumers such as ``ai/1.0``, ``ai/1.1``, ``ai/1.2`` which in turn each one might be used by multiple ``engine`` different versions and so on. Building only the latest version of the consumers would be insufficient in many cases and building all of  them would be extremely costly.
+- The "inverse" dependency model, that is, asking what are the "dependants" of a given package is extremely challeging in practice, specially in a decentralized
+  approach like Conan in which packages can be stored in different repositories, including different servers, and there isn't a central database of all packages and their relations.
+  Also, the "inverse" dependency model is, similar to the direct one, conditional. As a dependency can be conditional on any configuration (settings, options), the inverse is
+  also conditioned to the same logic, and such logic also evolves and changes with every new revision and version.
+
+In C and C++ projects the "products" pipeline becomes more necessary and critical than in other languages due to the compilation model with headers textual inclusions becoming part of the consumers' binary artifacts and due to the native artifacts linkage models. 
+
```

Comment:
Thanks for the feedback!!

> Out of curiosity: how does a CCI fork fit into this picture? Does it essentially become just one more source of packages that then get promoted into products?

Our own CCI uses the exact same principles, tools and flows. There are some differences, to handle with the scale, the reporting, security, etc.

For users CCI forks, the flow might change a little bit, because it is likely that the main flow is to update, merging from upstream, and there might be several packages affected at the same time. For this flow, doing a batch ``conan export`` operation for all the intended packages, then doing the ``conan graph build-order`` for the different configurations, using the exported changes, that part would be basically the same. So mainly the initial part, as the tutorial focused on changes per-package that lives in different git repos, one repo per package, while conan-center-index updates will likely be multi-package, and in the same repo.

Reply:
> Do I, as a CI engineer, just talk to the Important People and then proclaim "game and mapviewer has been declared as the main and only products, and thus we have CI only for them"?
Or should this be a more dynamic and programmatic approach, in essence determining all the root nodes in the dependency graph?
Or must I not even think about traversing the graph at all, since I will inevitably fall victim to the same issue of conditional dependencies?

Almost any attempt to try to deduce dynamically the "products" will most likely be worse than explicit modeling of those. Note that for a start, organizations have packages distributed accross many repositories, some global to the org, some specific for teams, etc. Packages often count in many thousands of them, each one with multiple versions and revisions. It becomes intractable pretty easily. 
Also, any automatic approach is completely unable to capture any business logic and decision.

So mostly, it is a explicit definition, not by the CI engineer, but it comes from the business: This org/team is delivering these 3 products, ``app1``, ``app2`` and ``app3``. For each one of them we might need to maintain different versions, as we have customers that depend on them. So our products are ``app1/2.3``, ``app1/3.8``, ``app2/1.1`` and ``app3/1.2``, ``app3/1.5``. When our customers upgrade we can safely remove some of these versions from our products, and we can also incorporate new versions of the products for the forecoming new releases. 

These products are the things that shouldn't be broken when we create new versions of the packages. For legacy products no longer in our current product list, we will be able to re-build them, if we captured the recommended lockfiles when they were last built, but if we don't use the lockfile to constraint new dependencies versions, they might indeed break if trying to use latest versions. This is expected and desired, we didn't want to burn CI for a product that is no longer maintained and evolved!